### ISSUE TITLE
Failed Build branch: linux

### ISSUE THREAD

**Issue by TomSchober** [OpenSpace/OpenSpace#1167]
@ e6b44b5b4c1945ae92fe5e4e0f7125bf24307467

Following [build instructions](http://wiki.openspaceproject.com/docs/developers/compiling/ubuntu)

Testing this in a VM:
```
vmName="$1"
downloadDir="$2"

downloadUrl="https://releases.ubuntu.com/18.04.4/ubuntu-18.04.4-desktop-amd64.iso"
downloadTarget="${downloadDir}/ubuntu-18.04.4-desktop-amd64.iso"
checksum=c0d025e560d54434a925b3707f8686a7f588c42a5fbc609b8ea2447f88847041

if [[ ! -f "${downloadTarget}" ]]; then
  wget "${downloadUrl}" -O "${downloadTarget}"
fi

echo "Checking ISO..."
if [ ! -f "${downloadTarget}" ];
then
  error "File $downloadTarget doesn't exist"
elif [ ! "$(sha256sum "$downloadTarget" | head -c 64)" = "$checksum" ]; then
  echo "Checksum failed" >&2
  exit 1
fi

vboxmanage createvm --name "${vmName}" --register
vboxmanage modifyvm "${vmName}" \
--ostype Ubuntu --memory 8192 \
--cpus 4 --memory 16384 --vram 128 \
--accelerate3d on --graphicscontroller vmsvga --audio alsa --audiocontroller ac97 \
--nic1 bridged --bridgeadapter1 enp8s0 --nictype1 82545EM --cableconnected1 on

vboxmanage createhd --filename "$HOME/VirtualBox VMs/$vmName/$vmName.vdi" --size 51200 --format VDI
vboxmanage storagectl "${vmName}" --name SATA --add sata --controller IntelAhci --bootable on
vboxmanage storageattach "${vmName}" --storagectl SATA --port 0 --device 0 --type hdd --medium "$HOME/VirtualBox VMs/$vmName/$vmName.vdi"
vboxmanage storagectl "${vmName}" --name IDE --add ide --controller PIIX4 --bootable on
vboxmanage storageattach "${vmName}" --storagectl IDE --port 1 --device 0 --type dvddrive --medium ${downloadTarget}

vboxmanage sharedfolder add "${vmName}" --name openSpace --hostpath "${HOME}/source/other/OpenSpace" --automount

vboxmanage startvm "${vmName}"
```

1. Intstall Ubuntu Desktop 18.04.4 (graphical / minimal install)
1. Install Guest additions to assist in sharing
1. reboot

With some small modifications to the Wiki instructions on getting started  run this in home directory:
(not my recommendation on a bash script, just keeping in line with the wiki)
```
sudo apt-get update
sudo apt-get -y upgrade
sudo apt-get -y dist-upgrade
sudo apt-get install -y build-essential software-properties-common
sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
sudo apt-get update
sudo apt-get install -y gcc-8 g++-8
sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-8 60 --slave /usr/bin/g++ g++ /usr/bin/g++-8
sudo update-alternatives --config gcc

sudo apt-get install -y glew-utils
sudo apt-get install -y freeglut3-dev
sudo apt-get install -y libxrandr-dev \
libxinerama-dev \
xorg-dev \
libcurl4-openssl-dev \
libgdal-dev \
libxcursor-dev

# Missing cmake in instructions
sudo apt-get install -y cmake

sudo apt-get install -y vim git

git clone --recursive --branch linux https://github.com/OpenSpace/OpenSpace "$HOME/OpenSpace"
mkdir "$HOME/OpenSpace/build"
cd "$HOME/OpenSpace/build" || exit
cmake ..
make
```

Result:
Fails near then end of compilation with
```
[ 98%] Linking CXX static library libvrpn.a
[ 98%] Built target vrpn
Scanning dependencies of target sgct
[ 98%] Building CXX object apps/OpenSpace/ext/sgct/src/sgct/CMakeFiles/sgct.dir/baseviewport.cpp.o
cc1plus: warning: /home/openspace/OpenSpace/build/apps/OpenSpace/ext/sgct/ext/zlib/zconf.h: not a directory
/home/openspace/OpenSpace/apps/OpenSpace/ext/sgct/src/sgct/baseviewport.cpp:17:10: fatal error: glm/glm.hpp: No such file or directory
 #include <glm/glm.hpp>
          ^~~~~~~~~~~~~
compilation terminated.
apps/OpenSpace/ext/sgct/src/sgct/CMakeFiles/sgct.dir/build.make:62: recipe for target 'apps/OpenSpace/ext/sgct/src/sgct/CMakeFiles/sgct.dir/baseviewport.cpp.o' failed
make[2]: *** [apps/OpenSpace/ext/sgct/src/sgct/CMakeFiles/sgct.dir/baseviewport.cpp.o] Error 1
CMakeFiles/Makefile2:3036: recipe for target 'apps/OpenSpace/ext/sgct/src/sgct/CMakeFiles/sgct.dir/all' failed
make[1]: *** [apps/OpenSpace/ext/sgct/src/sgct/CMakeFiles/sgct.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by TomSchober** [comment]
@GPayne You look to be the most recent active developer in the `linux` branch.  Any insight you might have in the gaps between the documentation and current development practices would be very appreciedated.  --  Thanks

**Comment by GPayne** [comment]
@TomSchober  Thanks for working on this and creating the issue. Your bash script looks good, I think it is only missing the CMake variables from the wiki. So if you change your cmake command (second-to-last line) to:
`cmake -DCMAKE_CXX_COMPILER:FILEPATH=/usr/bin/g++-8 -DCMAKE_C_COMPILER:FILEPATH=/usr/bin/gcc-8 -DCMAKE_CXX_FLAGS:STRING="-std=gnu++17 -DGLM_ENABLE_EXPERIMENTAL" -DOpenGL_GL_PREFERENCE:STRING=GLVND ..`
then it will help.

**Comment by TomSchober** [comment]
@GPayne ... good advice... I had also tried with:
```
-DCMAKE_CXX_FLAGS:STRING=-std=gnu++17 -DGLM_ENABLE_EXPERIMENTAL:STRING=ON -DOPENGL_GL_PREFERENCE:STRING=GLVND
```
as elluded to [here](http://wiki.openspaceproject.com/docs/developers/compiling/ubuntu#gcc) with the same results.  If there is a README I am not seeing I can help clarify the build process for others inline (markup).

So now with my steps above, and altering `cmake` to this:
```
cmake -DCMAKE_CXX_COMPILER:FILEPATH=/usr/bin/g++-8 -DCMAKE_C_COMPILER:FILEPATH=/usr/bin/gcc-8 -DCMAKE_CXX_FLAGS:STRING="-std=gnu++17 -DGLM_ENABLE_EXPERIMENTAL" -DOpenGL_GL_PREFERENCE:STRING=GLVND ..

make
```

results in:
```
Scanning dependencies of target sgct
[ 98%] Building CXX object apps/OpenSpace/ext/sgct/src/sgct/CMakeFiles/sgct.dir/baseviewport.cpp.o
cc1plus: warning: /home/openspace/OpenSpace/build/apps/OpenSpace/ext/sgct/ext/zlib/zconf.h: not a directory
/home/openspace/OpenSpace/apps/OpenSpace/ext/sgct/src/sgct/baseviewport.cpp:17:10: fatal error: glm/glm.hpp: No such file or directory
 #include <glm/glm.hpp>
          ^~~~~~~~~~~~~
compilation terminated.
apps/OpenSpace/ext/sgct/src/sgct/CMakeFiles/sgct.dir/build.make:62: recipe for target 'apps/OpenSpace/ext/sgct/src/sgct/CMakeFiles/sgct.dir/baseviewport.cpp.o' failed
make[2]: *** [apps/OpenSpace/ext/sgct/src/sgct/CMakeFiles/sgct.dir/baseviewport.cpp.o] Error 1
CMakeFiles/Makefile2:3036: recipe for target 'apps/OpenSpace/ext/sgct/src/sgct/CMakeFiles/sgct.dir/all' failed
make[1]: *** [apps/OpenSpace/ext/sgct/src/sgct/CMakeFiles/sgct.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2

```

I do see a Jenkinsfile.  Is jenkins set up for a multi-branch pipeline?  This way we could see when a regression may have occurred in the `linux` branch.

**Comment by GPayne** [comment]
@TomSchober : big changes are coming soon. A colleague found the cause of the rendering problem in the sgct submodule, and will be pushing his changes soon. I suggest waiting for the linux branch changes (which will later be merged into master). I will post an update here when the changes get pushed.

**Comment by TomSchober** [comment]
I discovered the root cause: The instructions are missing a dependency in [the docs](http://wiki.openspaceproject.com/docs/developers/compiling/ubuntu.html).
```
libglm-dev
```
I updated the test script and dependency install, in a [Gist](https://gist.github.com/TomSchober/987f2ffbe92f4b0d4279091d98bf667a):
```shell
# Prerequisite: VirtualBox >= 6.1.6 with extension pack
# Example: ./vboxSetup.sh "OpenSpace DEV1" "/home/user/Downloads"

vmName="$1"
downloadDir="$2"

downloadUrl="https://releases.ubuntu.com/18.04.4/ubuntu-18.04.4-desktop-amd64.iso"
downloadTarget="${downloadDir}/ubuntu-18.04.4-desktop-amd64.iso"
checksum=c0d025e560d54434a925b3707f8686a7f588c42a5fbc609b8ea2447f88847041

if [[ ! -f "${downloadTarget}" ]]; then
  wget "${downloadUrl}" -O "${downloadTarget}"
fi

echo "Checking ISO..."
if [ ! -f "${downloadTarget}" ];
then
  error "File $downloadTarget doesn't exist"
elif [ ! "$(sha256sum "$downloadTarget" | head -c 64)" = "$checksum" ]; then
  echo "Checksum failed" >&2
  exit 1
fi

vboxmanage createvm --name "${vmName}" --register
vboxmanage modifyvm "${vmName}" \
--ostype Ubuntu --memory 8192 \
--cpus 4 --memory 16384 --vram 128 \
--accelerate3d on --graphicscontroller vmsvga --audio alsa --audiocontroller ac97 \
--nic1 bridged --bridgeadapter1 enp8s0 --nictype1 82545EM --cableconnected1 on \
--clipboard-mode bidirectional --draganddrop bidirectional

vboxmanage createhd --filename "$HOME/VirtualBox VMs/$vmName/$vmName.vdi" --size 51200 --format VDI
vboxmanage storagectl "${vmName}" --name SATA --add sata --controller IntelAhci --bootable on
vboxmanage storageattach "${vmName}" --storagectl SATA --port 0 --device 0 --type hdd --medium "$HOME/VirtualBox VMs/$vmName/$vmName.vdi"
vboxmanage storagectl "${vmName}" --name IDE --add ide --controller PIIX4 --bootable on
vboxmanage storageattach "${vmName}" --storagectl IDE --port 1 --device 0 --type dvddrive --medium ${downloadTarget}

vboxmanage sharedfolder add "${vmName}" --name source --hostpath "${HOME}/source" --automount

vboxmanage startvm "${vmName}"

```

```shell
# Prerequisite: Ubuntu 18.04.4 Desktop (minimal installation with VirtualBox Guest Additions)

# Dependencies
sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
sudo apt-get update
sudo apt-get install -y \
build-essential software-properties-common \
gcc-8 g++-8 cmake \
glew-utils freeglut3-dev libsoil1 \
libxrandr-dev libxinerama-dev xorg-dev libcurl4-openssl-dev libgdal-dev libxcursor-dev \
libglm-dev \
vim git

# Build
openSpaceHome="$HOME/OpenSpace"

git clone --recursive --branch linux https://github.com/OpenSpace/OpenSpace "$openSpaceHome"
mkdir "$openSpaceHome/build"
cd "$openSpaceHome/build" || exit
cmake \
-DCMAKE_CXX_COMPILER:FILEPATH=/usr/bin/g++-8 \
-DCMAKE_C_COMPILER:FILEPATH=/usr/bin/gcc-8 \
-DCMAKE_CXX_FLAGS:STRING="-std=gnu++17 -DGLM_ENABLE_EXPERIMENTAL" \
-DOpenGL_GL_PREFERENCE:STRING=GLVND "$openSpaceHome"
make
```

@GPayne Where might I help update the documentation to help others?

This doesn't fix all of the issues, but I will wait for the updated [linux-fixes](https://github.com/OpenSpace/OpenSpace/tree/linux-fixes) branch by @jccosta before digging any further :

```shell
openspace@openspace-VirtualBox:~/OpenSpace/bin$ export MESA_GL_VERSION_OVERRIDE=4.3
openspace@openspace-VirtualBox:~/OpenSpace/bin$ ./OpenSpace
```
```
TransformationManager (Warning) : Kameleon module needed for transformations with dynamic frames
Lua (Debug) : Creating Lua state
Lua (Debug) : Open libraries
Lua (Debug) : Creating Lua state
Lua (Debug) : Open libraries
FileSystem (Info)	Set temporary path ${TEMPORARY} to /tmp
main (Debug)	Finding configuration
main (Info)	Configuration Path: '/home/openspace/OpenSpace/openspace.cfg'
main (Debug)	Loading configuration from disk
Lua (Debug)	Creating Lua state
Lua (Debug)	Open libraries
main (Debug)	SGCT Configuration file: /tmp/lua_6efWA6
OpenSpaceEngine (Debug)	Registering path ${ASSETS}: ${DATA}/assets
OpenSpaceEngine (Debug)	Registering path ${CACHE}: ${BASE}/cache
OpenSpaceEngine (Debug)	Registering path ${CONFIG}: ${BASE}/config
OpenSpaceEngine (Debug)	Registering path ${DATA}: ${BASE}/data
OpenSpaceEngine (Debug)	Registering path ${DOCUMENTATION}: ${BASE}/documentation
OpenSpaceEngine (Debug)	Registering path ${FONTS}: ${DATA}/fonts
OpenSpaceEngine (Debug)	Registering path ${LOGS}: ${BASE}/logs
OpenSpaceEngine (Debug)	Registering path ${MODULES}: ${BASE}/modules
OpenSpaceEngine (Debug)	Registering path ${RECORDINGS}: ${BASE}/recordings
OpenSpaceEngine (Debug)	Registering path ${SCREENSHOTS}: ${BASE}/screenshots
OpenSpaceEngine (Debug)	Registering path ${SCRIPTS}: ${BASE}/scripts
OpenSpaceEngine (Debug)	Registering path ${SHADERS}: ${BASE}/shaders
OpenSpaceEngine (Debug)	Registering path ${SYNC}: ${BASE}/sync
OpenSpaceEngine (Debug)	Registering path ${TASKS}: ${DATA}/tasks
OpenSpaceEngine (Debug)	Registering path ${TEMPORARY}: ${BASE}/temp
OpenSpaceEngine (Debug)	Registering path ${WEB}: ${DATA}/web
main (Debug)	Creating SGCT Engine
SGCT (Info)	Number of nodes in cluster: 
SGCT (Info)		Node(0) address: localhost [20401
SGCT (Info)	SGCT version: 3.0.
SGCT (Debug)	Validating cluster configuratio
SGCT (Debug)	Initiating network AP
SGCT (Debug)	Getting host inf
SGCT (Debug)	Matching ip address to find node in configuratio
SGCT (Debug)	Running in cluster mode as node 
SGCT (Info)	This computer is the network serve
SGCT (Debug)	Initiating connection 0 at port 2050
SGCT (Debug)	Cluster sync: loos
SGCT (Info)	Waiting for client 0 to connect on port 20500
SGCT (Info)	Detected OpenGL version: 4.
SGCT (Info)	Using GLFW version 3.3.
SGCT (Info)	OpenGL version 4.3.0 core profil
SGCT (Info)	Vendor: VMware, Inc
SGCT (Info)	Renderer: SVGA3D; build: RELEASE;  LLVM
SGCT (Error)	Vertex shader compile error: 0:2(10): error: GLSL 3.30 is not supported. Supported versions are: 1.10, 1.20, and 1.00 ES
SGCT (Error)	Fragment shader compile error: 0:2(10): error: GLSL 3.30 is not supported. Supported versions are: 1.10, 1.20, and 1.00 ES
SGCT (Error)	Shader [FBOQuadShader] linking error: error: linking with uncompiled/unspecialized shadererror: linking with uncompiled/unspecialized shade
SGCT (Info)	Cleaning u
SGCT (Debug)	Clearing callback
SGCT (Info)	Deleting screen capture data for window 
SGCT (Info)	Releasing OpenGL buffers for window 
SGCT (Info)	Deleting VBOs for window 
SGCT (Info)	Deleting VAOs for window 
SGCT (Debug)	Destroying network manage
SGCT (Info)	Closing connection 
SGCT (Info)	Exiting connection handler for connection 
SGCT (Error)	Accept connection 0 failed. Error: 2
SGCT (Debug)	Updating status for connection 
SGCT (Info)	Number of active connections 0 of 
SGCT (Debug)	Number of connected sync nodes 0 of 
SGCT (Debug)	Number of connected data transfer nodes 0 of 
SGCT (Error)	[Network] (5015): Send data failed: 
SGCT (Info)	Connection 0 successfully terminate
SGCT (Info)	Connection 0 successfully terminate
SGCT (Info)	Network API close
SGCT (Debug)	Destroying shader manager and internal shader
SGCT (Debug)	Destroying texture manage
SGCT (Debug)	Destroying font manage
SGCT (Debug)	Destroying shared dat
SGCT (Debug)	Destroying cluster manage
SGCT (Debug)	Destroying setting
SGCT (Debug)	Destroying message handle
ModuleEngine (Debug)	Deinitializing modules
ModuleEngine (Debug)	Finished deinitializing modules
ModuleEngine (Debug)	Finished destroying modules
Segmentation fault (core dumped)
```

**Comment by GPayne** [comment]
Nice work @TomSchober, and thanks for helping with this. I saw your wiki pull request and will review that.

Adding the `libglm-dev` isn't what we want to do, because we included glm code in our repository. I made a change to FindGLM.cmake so it finds the files there rather than the installed system version.

I am working on getting @jccosta 's linux-fixes branch (which he did on his Ubuntu 20.04 system) working on my 18.04 machine. I'm still getting a segfault that I need to figure out.

**Comment by TomSchober** [comment]
Understood.... I'll remove the package install from my update.

**Comment by jichengtao** [comment]
> @TomSchober : big changes are coming soon. A colleague found the cause of the rendering problem in the sgct submodule, and will be pushing his changes soon. I suggest waiting for the linux branch changes (which will later be merged into master). I will post an update here when the changes get pushed.

Hi, when the OpenSapce will work on Ubuntu?

**Comment by jccosta** [comment]
Hi @jichengtao and @TomSchober, I made some changes in OS to compile under Linux. The new changes are already in the master branch.
You will need to compile it using the release version; otherwise, you can have some issues (-DCMAKE_BUILD_TYPE=Release if you are using the command-line options).
In OpenSpace Support in Slack (https://join.slack.com/t/openspacesupport/shared_invite/enQtMjUxNzUyMTQ1ODQxLTRmNDI1YTA4ODkzODUyODE0YjIzODU0NWU1NGY1NWIzZDUzMDgwM2VkYmE1ZGY3MmU2OWI5NzhlN2U3NWU2NTQ), there is a #linux channel with an active conversation about how to compile OS on Linux using GCC or clang.
Please let me know if you have success compiling OS on Linux.

**Comment by TomSchober** [comment]
@jichengtao & @jccosta I'm closing this because the original problem was resolved.  We can continue in another issue if you like.

**Comment by jichengtao** [comment]
> Hi @jichengtao and @TomSchober, I made some changes in OS to compile under Linux. The new changes are already in the master branch.
> You will need to compile it using the release version; otherwise, you can have some issues (-DCMAKE_BUILD_TYPE=Release if you are using the command-line options).
> In OpenSpace Support in Slack (https://join.slack.com/t/openspacesupport/shared_invite/enQtMjUxNzUyMTQ1ODQxLTRmNDI1YTA4ODkzODUyODE0YjIzODU0NWU1NGY1NWIzZDUzMDgwM2VkYmE1ZGY3MmU2OWI5NzhlN2U3NWU2NTQ), there is a #linux channel with an active conversation about how to compile OS on Linux using GCC or clang.
> Please let me know if you have success compiling OS on Linux.

Hi, @jccosta , I tried the linux-fixes branch, and it compiled successfully under the Ubuntu 20.04. But while running, I still got the black screen.

**Comment by jccosta** [comment]
You must compile it in release mode: -DCMAKE_BUILD_TYPE=Release.

**Comment by jccosta** [comment]
Also, you must use the master branch.

**Comment by jichengtao** [comment]
> Also, you must use the master branch.

Hi @jccosta , thanks a lot, compiled successfully and run well under Ubuntu 20.04 now.

**Comment by jccosta** [comment]
@jichengtao Good to hear it.

**Comment by gitkime** [comment]
Hi @TomSchober 
I tried to run OpenSpace on Ubuntu 18.04. I followed the issue, but I don't understand, if OpenSpace is now running on Ubuntu 18.04, or have i switch to Ubuntu 20.04?
Can I clone the master repo and try to compile? Thanks for you hints.
@gitkime

**Comment by GPayne** [comment]
Hi @gitkime, I was able to build & run OpenSpace on 18.04. It should work for you if you follow the updated directions at http://wiki.openspaceproject.com/docs/developers/compiling/ubuntu.html.
